### ISSUE TITLE
twisted.python.rebuild.rebuild dies if something's __hash__ raises an exception other than TypeError

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3817]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| z3p reported|
|-|-|
|Trac ID|trac#3817|
|Type|defect|
|Created|2009-05-04 15:50:20Z|
|Branch|https://github.com/twisted/twisted/tree/rebuild-hash-fail-3817|

This came up in [#3391](https://github.com/twisted/twisted/issues/3391), where PyASN1 causes this to fail test_rebuild.

<details><summary>Searchable metadata</summary>

```
trac-id__3817 3817
type__defect defect
reporter__z3p z3p
priority__normal normal
milestone__ 
branch__branches_rebuild_hash_fail_3817 branches/rebuild-hash-fail-3817
branch_author__z3p z3p
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1241452220000000 1241452220000000
changetime__1241988211000000 1241988211000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|z3p commented|
|-|-|

(In [26824]) Branching to 'rebuild-hash-fail-3817'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|z3p commented|
|-|-|

Did I say 'enhancement'? I meant 'defect'.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|z3p removed owner|
|-|-|

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to z3p<br>@exarkun set status to `new`|
|-|-|

  1. twisted/test/test_rebuild.py
    1. name all new test methods `test_foo` instead of `testFoo`
    1. use two blank lines between new or modified methods and the method preceding or following; use three between a method and the start of the next class
    1. It might be nice to avoid adding an unhashable attribute to `test_rebuild` permanently (in case some other tool can't handle this case).  The test could set the attribute when it starts and use `addCleanup` to make sure it is deleted when the test is over.  (this is an optional review comment)

Please merge when all points are addressed.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|z3p set status to `closed`|
|-|-|

(In [26910]) Don't fail the rebuild if an object cannot be hashed.

Author: z3p
Reviewer: exarkun
Fixes [#3817](https://github.com/twisted/twisted/issues/3817)

Previously, only a TypeError was ignored for objects during a module rebuild.  Some objects raise other errors (notably the PyASN1 module) and caused the test_rebuild tests to fail.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
